File: src/main/java/com/healthpal/api/service/AuthService.java
package com.healthpal.api.service;

import com.healthpal.api.model.User;
import com.healthpal.api.model.Role;
import com.healthpal.api.model.dto.SignUpRequest;
import com.healthpal.api.model.dto.LoginRequest;
import com.healthpal.api.model.dto.AuthResponse;
import com.healthpal.api.repository.UserRepository;
import com.healthpal.api.repository.RoleRepository;
import com.healthpal.api.util.JwtTokenProvider;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * AuthService - Business Logic Layer for Authentication
 * 
 * This service is part of the MVC Layered Architecture (Service/Business Logic Layer)
 * It handles:
 * - User registration (sign up)
 * - User login
 * - Token generation and validation
 * - User role assignment
 * 
 * @author HealthPal Team - Person A (Authentication + Admin + Core Architecture)
 * @version 1.0
 */
@Slf4j
@Service
@Transactional
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private RoleRepository roleRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    @Autowired
    private AuthenticationManager authenticationManager;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    /**
     * Register a new user
     * @param signUpRequest The registration request containing user details
     * @return AuthResponse with success status and message
     * @throws RuntimeException if username or email already exists, or passwords don't match
     */
    public AuthResponse signUp(SignUpRequest signUpRequest) {
        log.info("Attempting to register user with username: {}", signUpRequest.getUsername());

        // Validate that passwords match
        if (!signUpRequest.getPassword().equals(signUpRequest.getConfirmPassword())) {
            log.warn("Password mismatch for user: {}", signUpRequest.getUsername());
            throw new RuntimeException("Passwords do not match");
        }

        // Check if username already exists
        if (userRepository.existsByUsername(signUpRequest.getUsername())) {
            log.warn("Username already exists: {}", signUpRequest.getUsername());
            throw new RuntimeException("Username is already taken");
        }

        // Check if email already exists
        if (userRepository.existsByEmail(signUpRequest.getEmail())) {
            log.warn("Email already exists: {}", signUpRequest.getEmail());
            throw new RuntimeException("Email is already registered");
        }

        // Create new user
        User user = User.builder()
                .username(signUpRequest.getUsername())
                .email(signUpRequest.getEmail())
                .password(passwordEncoder.encode(signUpRequest.getPassword()))
                .firstName(signUpRequest.getFirstName())
                .lastName(signUpRequest.getLastName())
                .phoneNumber(signUpRequest.getPhoneNumber())
                .isActive(true)
                .isEmailVerified(false)
                .isPhoneVerified(false)
                .isAdmin(false)
                .build();

        // Assign default role based on user type
        String userType = signUpRequest.getUserType().toUpperCase();
        String roleName = "ROLE_" + userType;

        Role role = roleRepository.findByName(roleName)
                .orElseThrow(() -> new RuntimeException("Role not found: " + roleName));

        user.addRole(role);

        // Save user to database
        User savedUser = userRepository.save(user);
        log.info("User registered successfully: {}", savedUser.getUsername());

        return AuthResponse.builder()
                .userId(savedUser.getId())
                .username(savedUser.getUsername())
                .email(savedUser.getEmail())
                .success(true)
                .message("User registered successfully. Please verify your email.")
                .build();
    }

    /**
     * Authenticate user and generate JWT token
     * @param loginRequest The login request containing username/email and password
     * @return AuthResponse with JWT token and user information
     * @throws RuntimeException if authentication fails
     */
    public AuthResponse login(LoginRequest loginRequest) {
        log.info("Attempting to login user: {}", loginRequest.getUsernameOrEmail());

        try {
            // Authenticate user
            Authentication authentication = authenticationManager.authenticate(
                    new UsernamePasswordAuthenticationToken(
                            loginRequest.getUsernameOrEmail(),
                            loginRequest.getPassword()
                    )
            );

            // Set authentication in security context
            SecurityContextHolder.getContext().setAuthentication(authentication);

            // Generate JWT token
            String token = jwtTokenProvider.generateToken(authentication);
            String refreshToken = jwtTokenProvider.generateRefreshToken(authentication.getName());

            // Get user details
            User user = userRepository.findByUsernameOrEmail(
                    loginRequest.getUsernameOrEmail(),
                    loginRequest.getUsernameOrEmail()
            ).orElseThrow(() -> new RuntimeException("User not found"));

            // Update last login time
            user.setLastLoginAt(java.time.LocalDateTime.now());
            userRepository.save(user);

            // Extract roles
            Set<String> roles = user.getRoles().stream()
                    .map(Role::getName)
                    .collect(Collectors.toSet());

            log.info("User logged in successfully: {}", user.getUsername());

            return AuthResponse.builder()
                    .token(token)
                    .refreshToken(refreshToken)
                    .userId(user.getId())
                    .username(user.getUsername())
                    .email(user.getEmail())
                    .firstName(user.getFirstName())
                    .lastName(user.getLastName())
                    .isAdmin(user.getIsAdmin())
                    .roles(roles)
                    .success(true)
                    .message("Login successful")
                    .build();

        } catch (Exception e) {
            log.error("Login failed for user: {}", loginRequest.getUsernameOrEmail(), e);
            throw new RuntimeException("Invalid username or password");
        }
    }

    /**
     * Refresh JWT token
     * @param refreshToken The refresh token
     * @return AuthResponse with new JWT token
     * @throws RuntimeException if refresh token is invalid
     */
    public AuthResponse refreshToken(String refreshToken) {
        log.info("Attempting to refresh token");

        if (!jwtTokenProvider.validateToken(refreshToken)) {
            log.warn("Invalid refresh token");
            throw new RuntimeException("Invalid refresh token");
        }

        String username = jwtTokenProvider.getUsernameFromToken(refreshToken);
        String newToken = jwtTokenProvider.generateTokenFromUsername(username);

        User user = userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("User not found"));

        Set<String> roles = user.getRoles().stream()
                .map(Role::getName)
                .collect(Collectors.toSet());

        log.info("Token refreshed successfully for user: {}", username);

        return AuthResponse.builder()
                .token(newToken)
                .userId(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .roles(roles)
                .success(true)
                .message("Token refreshed successfully")
                .build();
    }

    /**
     * Get current authenticated user
     * @return User object of the current authenticated user
     * @throws RuntimeException if no user is authenticated
     */
    public User getCurrentUser() {
        String username = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByUsername(username)
                .orElseThrow(() -> new RuntimeException("Current user not found"));
    }
}

File: service/authService.js
const userRepository = require('../repositories/userRepository');
const roleRepository = require('../repositories/roleRepository');
const jwtTokenProvider = require('../utils/jwtTokenProvider');
const bcrypt = require('bcryptjs');
const logger = require('../utils/logger');
const AuthResponse = require('../dtos/AuthResponse');

class AuthService {
    async signUp(signUpRequest) {
        try {
            logger.info(`Attempting to register user with username: ${signUpRequest.username}`);
            if (signUpRequest.password !== signUpRequest.confirmPassword) {
                logger.warn(`Password mismatch for user: ${signUpRequest.username}`);
                throw new Error('Passwords do not match');
            }

            const usernameExists = await userRepository.existsByUsername(signUpRequest.username);
            if (usernameExists) {
                logger.warn(`Username already exists: ${signUpRequest.username}`);
                throw new Error('Username is already taken');
            }

            const emailExists = await userRepository.existsByEmail(signUpRequest.email);
            if (emailExists) {
                logger.warn(`Email already exists: ${signUpRequest.email}`);
                throw new Error('Email is already registered');
            }

            const userType = signUpRequest.userType.toUpperCase();
            const roleName = `ROLE_${userType}`;
            const role = await roleRepository.findByName(roleName);
            if (!role) {
                throw new Error(`Role not found: ${roleName}`);
            }

            const hashedPassword = await bcrypt.hash(signUpRequest.password, 10);
            const userData = {
                username: signUpRequest.username,
                email: signUpRequest.email,
                password: hashedPassword,
                firstName: signUpRequest.firstName,
                lastName: signUpRequest.lastName,
                phoneNumber: signUpRequest.phoneNumber,
                isActive: true,
                isEmailVerified: false,
                isPhoneVerified: false,
                isAdmin: userType === 'ADMIN',
                roles: [role._id]
            };

            const savedUser = await userRepository.save(userData);
            logger.info(`User registered successfully: ${savedUser.username}`);

            return AuthResponse.success({
                userId: savedUser._id,
                username: savedUser.username,
                email: savedUser.email,
                message: 'User registered successfully. Please verify your email.'
            });
        } catch (error) {
            logger.error(`Sign up error: ${error.message}`, error);
            throw error;
        }
    }

    async login(loginRequest) {
        try {
            logger.info(`Attempting to login user: ${loginRequest.usernameOrEmail}`);
            const user = await userRepository.findByUsernameOrEmail(
                loginRequest.usernameOrEmail,
                loginRequest.usernameOrEmail
            );

            if (!user) {
                logger.warn(`User not found: ${loginRequest.usernameOrEmail}`);
                throw new Error('Invalid username/email or password');
            }

            if (!user.isActive) {
                logger.warn(`User account is inactive: ${loginRequest.usernameOrEmail}`);
                throw new Error('User account is inactive');
            }

            const isPasswordValid = await bcrypt.compare(loginRequest.password, user.password);
            if (!isPasswordValid) {
                logger.warn(`Invalid password for user: ${loginRequest.usernameOrEmail}`);
                throw new Error('Invalid username/email or password');
            }

            const tokenData = jwtTokenProvider.generateTokenPair(user);
            await userRepository.update(user._id, { lastLoginAt: new Date() });
            const roles = user.roles ? user.roles.map(role => role.name) : [];
            logger.info(`User logged in successfully: ${user.username}`);

            return AuthResponse.fromUserAndToken(user, tokenData);
        } catch (error) {
            logger.error(`Login failed for user: ${loginRequest.usernameOrEmail}`, error);
            throw new Error('Invalid username or password');
        }
    }

    async refreshToken(refreshToken) {
        try {
            logger.info('Attempting to refresh token');
            const decoded = jwtTokenProvider.validateToken(refreshToken);
            if (!decoded || decoded.tokenType !== 'refresh') {
                logger.warn('Invalid refresh token');
                throw new Error('Invalid refresh token');
            }

            const username = jwtTokenProvider.getUsernameFromToken(refreshToken);
            if (!username) {
                throw new Error('Invalid refresh token');
            }

            const user = await userRepository.findByUsername(username);
            if (!user) {
                throw new Error('User not found');
            }

            const newToken = jwtTokenProvider.generateTokenFromUsername(username);
            const newRefreshToken = jwtTokenProvider.generateRefreshToken(user);
            const roles = user.roles ? user.roles.map(role => role.name) : [];
            logger.info(`Token refreshed successfully for user: ${username}`);

            return AuthResponse.success({
                token: newToken,
                refreshToken: newRefreshToken,
                userId: user._id,
                username: user.username,
                email: user.email,
                roles: roles,
                message: 'Token refreshed successfully'
            });
        } catch (error) {
            logger.error(`Token refresh failed: ${error.message}`, error);
            throw new Error('Invalid refresh token');
        }
    }

    async getCurrentUser(userId) {
        try {
            const user = await userRepository.findById(userId);
            if (!user) {
                throw new Error('Current user not found');
            }
            return user;
        } catch (error) {
            logger.error(`Get current user error: ${error.message}`, error);
            throw error;
        }
    }

    async authenticate(usernameOrEmail, password) {
        try {
            const user = await userRepository.findByUsernameOrEmail(
                usernameOrEmail,
                usernameOrEmail
            );

            if (!user) {
                throw new Error('User not found');
            }

            const isPasswordValid = await bcrypt.compare(password, user.password);
            if (!isPasswordValid) {
                throw new Error('Invalid password');
            }

            if (!user.isActive) {
                throw new Error('User account is inactive');
            }

            return user;
        } catch (error) {
            logger.error(`Authentication error: ${error.message}`, error);
            throw error;
        }
    }

    async logout(userId) {
        try {
            logger.info(`Logging out user ID: ${userId}`);
            return { 
                success: true, 
                message: 'Logged out successfully' 
            };
        } catch (error) {
            logger.error(`Logout error: ${error.message}`, error);
            throw error;
        }
    }

    hasRequiredRoles(user, requiredRoles) {
        if (!user || !user.roles) {
            return false;
        }

        const userRoles = user.roles.map(role => role.name);
        return requiredRoles.some(role => userRoles.includes(role));
    }

    isAdmin(user) {
        return user && (user.isAdmin || this.hasRequiredRoles(user, ['ROLE_ADMIN']));
    }
}

module.exports = new AuthService();